### PR TITLE
Cache frame component names/units

### LINKED ISSUE #7030: Re-work of `SkyCoord` initializer
This has come up in several contexts, and is an important goal for improving coordinates performance.  The `SkyCoord` initializer has evolved "organically" since it's initial creation in v0.4, which arguably is a euphemism for "not well-planned".  @adrn and @mhvk and I have discussed a tentative plan for how to implement this with basically no direct impact to the end-user beyond (hopefully) significant performance improvements.

Also involved in this *might* be a way to tell SkyCoord "don't do so much checking" (see #7029 and #6941).  Philisophically this might seem a bit odd because one might think the initializer is the place for that, and `classmethod`s should be used for cleverer parsing.  But we're pretty much stuck with the current API, so that's basically a non-starter. Instead we *hope* we can do some clever ordering of the parsing so that simple/basic use cases are parsed quickly, bypassing other checks.  That lets us have our 🍰 and eat it too.  But we shall see once we actually dive in...  

Note that this is related to but not quite the same as #7029.  Also some related discussion in #6896,  #6941, and #5952.

### PR DESCRIPTION
**Pull request by eteq**
This extends the machinery developed in #7949 for caching frame attributes to also cache the component names and units of coordinate frames.

Note that this is *not* the factor of 2x or so speedup we had initially thought it might be.  I'm not sure if it was some other speedup in components used by these methods, or something else appearing as a new bottleneck, but I see only something in the realm of a 0-10% speedup here (in the scalar coordinate case) depending on exactly what `SkyCoord` or frame object input we are talking about.

That raises the questions of whether this is "worth it" - caches can be notoriously tricky to get right - I think I have here (#7949, which had a more measureable speedup, is essentially the same machinery), but I've certainly been wrong before... And for that reason I'm not milestoning 3.1, although if we agree there's no harm in this we can probably backport it since it's not really a "feature" per se (nor is it a bug fix... not sure what to call it, so I'm calling it "affects-dev" even though it isn't exactly a bug).

cc @adrn @mhvk @astrofrog

Oh, and I think we could say this closes #7030 in that I think it now hits all of the points that have been highlighted there... But maybe @adrn doesn't think so if it's still in the plan to build some more performant classmethods (which is one of the strategies mentioned in #7030)?

### CODE REVIEW

**Comment by pllim** [comment]
> caches can be notoriously tricky to get right 

For future maintainability, maybe a few in-code comments on what the new changes do in plain English?

**Comment by eteq** [comment]
Inspired by https://github.com/astropy/astropy/pull/8016#issuecomment-434038046 I added comments both here and in the similar machinery that #7949 instituted.

**Comment by eteq** [comment]
Note the failure *may* be real.  It looks to be in the asdf serialization machinery... @drdavella, do you have any idea what's going on here?

**Comment by eteq** [comment]
Pushing to 3.2 since it's not a blocker

**Comment by pllim** [comment]
@eteq , try rebasing. I remember @drdavella did some `asdf` extension thingy here not too long ago.

**Comment by bsipocz** [comment]
ping @eteq - is this on the plate for 3.2?

**Comment by adrn** [comment]
My feeling now is that it's not worth adding this if the speedup is modest - I think there are still places where we can get much more significant speedups without adding to the complexity of the cache, so I'm 👎 on merging!

**Comment by pllim** [comment]
Since there is no consensus yet, I moved the milestone. FYI.
